bin/ch-run.c

```diff
@@ -212,40 +274,38 @@ void fix_environment(struct args *args)
    }
 
    // --set-env and --unset-env.
-   for (int i = 0; args->env_deltas[i].action != END; i++) {
+   for (int i = 0; args->env_deltas[i].action != ENV_END; i++) {
       char *arg = args->env_deltas[i].arg;
-      if (args->env_deltas[i].action == SET_FILE) {
-         FILE *fp;
-         Tf (fp = fopen(arg, "r"), "--set-env: can't open: %s", arg);
-         for (int j = 1; true; j++) {
-            char *line = NULL;
-            size_t len = 0;
-            errno = 0;
-            if (-1 == getline(&line, &len, fp)) {
-               if (errno == 0)  // EOF
-                  break;
-               else             // error
-                  Tf (0, "--set-env: error reading: %s", arg);
+      if (args->env_deltas[i].action == ENV_SET) {  // --set-env
+         if (strchr(arg, '=') != NULL) {
+            // argument is variable name & value
+            envs_set((char *[]){ arg }, 1, NULL, args->c.env_expand);
+         } else {
+            // argument is filename
+            char **lines = NULL;
+            int line_ct = 0;
+            FILE *fp;
+            Tf (fp = fopen(arg, "r"), "--set-env: can't open: %s", arg);
+            for (line_ct = 0; true; line_ct++) {
+               char *line;
+               size_t line_len = 0;  // don't care but getline(3) must write
+               errno = 0;
+               if (-1 == getline(&line, &line_len, fp)) {
```

Comment:
Note the incredibly awkward interface for `getline(3)` error checking. This type of thing is one motivation for exceptions, but even in C this could have been done better. E.g. why not return zero on EOF? The only time I can think of that one could read zero bytes is the last line of the file if it doesn't have a newline, but in that case the previous line would be the last line. I think there must be some subtlety I'm missing.